### PR TITLE
Add `core release` command

### LINKED ISSUE #108: Add `core release list` and `core release get` commands
It would be useful to be able to easily query the available releases of WordPress core through the .org API and also get the download URLs for a given update.

A use case I'm currently working on that this would benefit is https://github.com/wp-cli/wp-cli-tests/issues/51

### PR DESCRIPTION
**Pull request by schlessera**
Adds retrieval of available releases from the wordpress.org API server.

Version constraints are interpreted in the following way right now:

- `4` => anything that starts with the major version 4. This would mean `>= 4.0.0 && < 5.0.0`.
- `4.8` => anything that starts with the major version 4 and minor version 8. This would mean `>= 4.8.0 && < 4.9.0`

Some notes:
- Patch, pre-release or build constraints are not supported yet, as the API server does not provide querying functionality for these.
- I initially added a `--locale` flag, but removed it again for now, as the API server does not seem to respect it for all results, only for the very latest version.

Fixes #108

### CODE REVIEW

**Comment by chesio** [comment]
I would suggest to update example output to include WordPress 5.2 release as well - with an actual change in required PHP version it would make the examples even more telling.

File: src/Core_Command.php [hunk 1]
```diff
@@ -1395,4 +1395,249 @@ function () use ( $new_zip_file ) {
...
+	 *
+	 * @param string $version Version string to parse.
+	 * @return array Associative array of version data.
+	 * @throws \WP_CLI\ExitException If the version string could not be parsed.
+	 */
+	public function get_version_data( $version ) {
+		$semver_regex = '/^v?(\d+)(?:\.(\d+))?(?:\.(\d+))?(?:-([0-9A-Z-.]+))?(?:\+([0-9A-Z-.]+)?)?$/i';
+
+		if ( ! preg_match( $semver_regex, $version, $matches ) ) {
+			WP_CLI::error(
```
Comment (thrijith) [inline src/Core_Command.php]: Same as above.

**Comment by thrijith** [comment]
Also as @chesio mentioned, an example of `5.2` would be a good.

**Comment by nylen** [comment]
Shouldn't this be called `releases` instead?

**Comment by schlessera** [comment]
> Shouldn't this be called `releases` instead?

I think the "resource" we're dealing with here is "a release", and you can fetch a list of multiples of this resource. Just as we have `plugin`, `post` or `option` as singular resource names.

For context, the main use case I have for this is to turn a generic "release to use for tests" into "the latest release matching this criteria to use for tests".

So, if I state that I want to run tests with `4.9`, I don't want to fetch `4.9.0`, but rather the latest bugfix release of the `4.9` branch.
(see https://github.com/wp-cli/wp-cli-tests/issues/51)

**Comment by schlessera** [comment]
BTW, this is still unmerged because the endpoint I'm using doesn't provide everything I need and I'm not sure I'm using the correct endpoint in the first place (see #108 for more details).

**Comment by nylen** [comment]
> I think the "resource" we're dealing with here is "a release", and you can fetch a list of multiples of this resource. Just as we have `plugin`, `post` or `option` as singular resource names.

I'm not sure the same principle applies - all the other examples have multiple subcommands, and in that case it should be `wp core release list`.

When I read `core release` I thought of `release` as a verb, and thought: "Why would someone use WP-CLI to do a core release?"

**Comment by nylen** [comment]
> For context, the main use case I have for this is to turn a generic "release to use for tests" into "the latest release matching this criteria to use for tests".

It seems like there are 2 uses cases then: listing all releases (optionally matching some criteria), or finding the latest release (optionally matching some criteria).

Tying back into the "resource" discussion, why not make this into a proper resource with appropriate subcommands for these cases? For example:

```
wp release list
wp release list --version=4.9
wp release latest
wp release latest --version=4.9
```

Using `--version` as a partial match also seems a bit off, since a version should point to a specific release. Maybe this should be `--branch` or `--series` instead?

**Comment by danielbachhuber** [comment]
Proceeding with https://github.com/wp-cli/wp-cli/issues/5594 for this repository. I've captured this PR to https://gist.github.com/danielbachhuber/c3f452c41ae22532df1123550fec5f84 in case this PR is auto-closed or broken in some way.

**Pull request closed without merging**